src/js/components/AppListItemComponent.jsx

```diff
@@ -19,6 +21,71 @@ var AppListItemComponent = React.createClass({
     model: React.PropTypes.object.isRequired
   },
 
+  getInitialState: function () {
+    return {numVisibleLabels: -1};
+  },
+
+  componentDidMount: function () {
+    // Avoid referencing window from Node context
+    if (global.window != null) {
+      window.addEventListener("resize", this.handleResize);
+      window.addEventListener("focus", this.handleResize);
+    }
+    this.updateNumVisibleLabels();
+
+  },
+
+  componentDidUpdate: function (prevProps) {
+    if (!_.isEqual(this.props, prevProps)) {
+      this.updateNumVisibleLabels();
+    }
+  },
+
+  componentWillUnmount: function () {
+    if (global.window != null) {
+      window.removeEventListener("resize", this.handleResize);
+      window.removeEventListener("focus", this.handleResize);
+    }
+  },
+
+  shouldComponentUpdate: function (nextProps, nextState) {
+    if (nextState.numVisibleLabels === -1 && !this.props.model.isGroup) {
+      this.updateNumVisibleLabels();
+      return false;
+    }
+    return !_.isEqual(this.props, nextProps) ||
+      !_.isEqual(this.state, nextState);
+  },
+
+  handleResize: function () {
+    requestAnimationFrame(this.updateNumVisibleLabels);
+  },
+
+  updateNumVisibleLabels: function () {
+    var labels = this.props.model.labels;
+    if (labels == null || Object.keys(labels).length === 0) {
+      return null;
+    }
+    let cellNode = this.getDOMNode().querySelector(".name-cell");
```

Comment:
I wasn't sure how to retrieve the list of labels, thus is used the ``querySelector``. If you have a minute to show me how to do so, I'm happy to refactor this using refs.